shared/subspace-data-retrieval/src/object_fetcher/partial_object.rs

```diff
@@ -0,0 +1,736 @@
+        if maybe_padding_data.is_empty() {
+            assert!(
+                suffix_data.is_empty(),
+                "suffix_data can only have bytes if there might be padding bytes"
+            );
+            return Self::new_without_padding(prefix_data, max_object_len, mapping);
+        }
+
+        trace!(
+            ?max_object_len,
+            ?mapping,
+            prefix_data = ?hex::encode(prefix_data),
+            maybe_padding_data = ?hex::encode(maybe_padding_data),
+            suffix_data = ?hex::encode(suffix_data),
+            "trying to create new partial object with padding",
+        );
+
+        // Count the padding bytes
+        let object_max_segment_padding_length = Self::count_padding_bytes(maybe_padding_data);
+
+        trace!(?object_max_segment_padding_length, "analysed padding");
+
+        // Keep the remaining non-padding and potential padding bytes
+        let mut prefix_data = prefix_data.to_vec();
+        prefix_data.extend_from_slice(maybe_padding_data);
+
+        // Find the possible lengths for the object, for each potential padding byte.
+        let mut lengths = BTreeSet::new();
+        let mut last_error = None;
+
+        for ignored_padding_length in 0..=object_max_segment_padding_length {
+            let length_prefix = padded_data(&prefix_data, ignored_padding_length, suffix_data)
+                .take(MAX_ENCODED_LENGTH_SIZE)
+                .collect::<Vec<u8>>();
+
+            match decode_data_length(length_prefix.as_slice(), max_object_len, mapping) {
+                // A valid length
+                Ok(Some((length_prefix_len, data_length))) => {
+                    lengths.insert(ObjectLength {
+                        data_length: length_prefix_len + data_length,
+                        ignored_padding_length,
+                    });
+                }
+                // Not enough data to decode all the lengths yet (this is a rare edge case).
+                Ok(None) => {
+                    if lengths.is_empty() {
+                        // We need more data to decode a length.
+                        return Ok(None);
+                    }
+
+                    // Optimisation: try any lengths we already have.
+                    let mut partial_object = Self {
+                        prefix_data,
+                        suffix_data: suffix_data.to_vec(),
+                        lengths,
+                    };
+
+                    if let Ok(Some(_)) = partial_object.try_reconstruct_object(mapping) {
+                        // The other lengths don't matter, because we've found the object.
+                        // Return the completed object to the caller, so it can find that object.
+                        return Ok(Some(partial_object));
+                    } else {
+                        // We need more data to decode the correct length.
+                        return Ok(None);
+                    }
```

Comment:
Delete this optimisation and see if the tests still pass. If they don't, document why it is needed.